mantidproject/mantid
# Fit Script Generator - Connect FunctionTreeView to data table

**Description of work.**
This PR follows on from #30068 . It connects the `FunctionTreeView` to the data table, and same with the check box above the `FunctionTreeView`. The following is now possible:

- Selecting individual or multiple datasets, and then add or remove functions to the `FunctionTreeView` for only these datasets by unticking the tick box.
- Add  or remove functions to all datasets by ticking the checkbox.
- Selecting between datasets will display the functions related to that dataset.
- Switching between Sequential and Simultaneous fitting will hide/show the global parameter column and multi domain index

It is also now possible to do the following:
- Editing parameter and attribute values
- Local and global ties
- constraints
- global parameters

The `EditLocalParameter` dialog will be added in a different PR.

Roughly half of this PR ~ 2000 lines are additions to unit tests including:
- [x] New `FitDomainTest` 
- [x] New `FitScriptGeneratorModelTest` 
- [x] System test for opening this interface (`FitScriptGeneratorStartupTest`)
- [x] Additions to `FitScriptGeneratorPresenterTest`
- [x] Additions to `FitScriptGeneratorViewTest`
- [x] Additions to `CompositeFunctionTest`

**To test:**
1. Comment out Line 52 in `fitting_tab_view.py`
1. `Muon Interface`
2. Load any data and go to fitting tab.
3. Click `Fit Generator` to open the new interface
4. Load more data using `Add Workspace` so that you have at least 2 rows in the data table.
5. Choose `Simultaneous` in the FitOptionsBrowser (bottom right). This should show the global column in the FunctionTreeView (top right), and the multi domain index on the left of the data table.
6. Add a FlatBackground for both datasets (keeping checkbox ticked).
7. Untick the checkbox, and select the first table row
8. Add an ExpDecay function.
9. Switch between the rows to make sure only the top row has the `ExpDecay`. Also notice that the function index at the top of the FunctionTreeView will change when switching between the table rows.
10. Remove the `ExpDecay`. Tick the checkbox. Add the `ExpDecay` function. It should now be there for every domain.
11. Try a local tie: Select the first row, and right click `f0.f1.Height` and tie it to `f0.f0.A0`. Remove the tie.
12. Try a global tie: Select the first row, and right click `f0.f1.Height` and tie it to `f1.f0.A0`. ie the domain index is different.
13. Remove this tie. Try ticking a parameter to make it global. Global parameters will not be allowed if:
 - The parameter does not exist in ALL domains
 - The parameter already has a local or global tie
14. Try a constraint

Closes #30198

#### Reviewer ####

Please comment on the following ([full description](http://developer.mantidproject.org/ReviewingAPullRequest.html)):

##### Code Review #####

- Is the code of an acceptable quality?
- Does the code conform to the [coding standards](http://developer.mantidproject.org/Standards/)?
- Are the unit tests small and test the class in isolation?
- If there is GUI work does it follow the [GUI standards](http://developer.mantidproject.org/Standards/GUIStandards.html)?
- If there are changes in the release notes then do they describe the changes appropriately?

##### Functional Tests #####

- Do changes function as described? Add comments below that describe the tests performed?
- Do the changes handle unexpected situations, e.g. bad input?
- Has the relevant (user and developer) documentation been added/updated?

Does everything look good? Mark the review as **Approve**. A member of `@mantidproject/gatekeepers` will take care of it.
